Update and improve docker and helm deployment docs (#1246)
zenml-io/zenml: docs/book/getting-started/deploying-zenml/helm.md

---
description: Deploying ZenML in a Kubernetes cluster with Helm
---

If you wish to manually deploy and manage ZenML in a Kubernetes cluster of your
choice, ZenML also includes a Helm chart among its available deployment options.

The Helm chart is not available on a public online repository yet, but it is
included in the ZenML source code repository. You can follow this guide to
deploy ZenML in any Kubernetes cluster using the Helm chart.

## Prerequisites

You'll need the following:

- A Kubernetes cluster.

> - Optional, but recommended: a MySQL compatible database reachable from the

Reviewer: ```suggestion
- Optional, but recommended: a MySQL-compatible database reachable from the
```

Kubernetes cluster (e.g. one of the managed databases offered by Google Cloud,
AWS or Azure). A MySQL server version of 8.0 or higher is required.

> - the [Kubernetes client](https://kubernetes.io/docs/tasks/tools/#kubectl)
> already installed on your machine and configured to access your cluster
> - [Helm](https://helm.sh/docs/intro/install/) installed on your machine

Reviewer: ```suggestion
- the [Kubernetes client](https://kubernetes.io/docs/tasks/tools/#kubectl)
already installed on your machine and configured to access your cluster.
- [Helm](https://helm.sh/docs/intro/install/) installed on your machine.
```

To gain access to the ZenML Helm chart, you'll need to clone the ZenML
repository and checkout the `main` branch, or one of the release tags:

```bash
git clone https://github.com/zenml-io/zenml.git
# Optional: checkout a previous release tag

> git checkout 0.21.1

Reviewer: ```suggestion
# git checkout 0.21.1 
```
If it's optional, it should probably be commented out?

# Switch to the directory that hosts the helm chart
cd src/zenml/zen_server/deploy/helm/
```

You can start by taking a look at the `values.yaml` file and familiarizing
yourself with some of the configuration settings that you can customize for your
ZenML deployment.

In addition to tools and infrastructure, you will also need to collect and
[prepare information related to your database](#collect-information-from-your-sql-database-service)
to be used for the Helm chart configuration and you may
also want to install additional [optional services in your cluster](#optional-cluster-services).

When you are ready, you can proceed to the [installation](#installation)
section.

### Collect information from your SQL database service

> Using an external MySQL compatible database service is optional, but is

Reviewer: ```suggestion
Using an external MySQL-compatible database service is optional, but is
```

recommended for production deployments. If omitted, ZenML will default to using
an embedded SQLite database, which has the following limitations:

> - the SQLite database is not persistent, meaning that it will be lost if the

Reviewer: ```suggestion
- the SQLite database is not persisted, meaning that it will be lost if the
```

cluster
- the username and password that will be used to connect to the database. It is
recommended that you create a dedicated database user for the ZenML server and
that you restrict its privileges to only access the database that will be used
by ZenML. Enforcing secure SSL connections for the user/database is also
recommended. See the [MySQL documentation](https://dev.mysql.com/doc/refman/5.7/en/access-control.html)
for more information on how to set up users and privileges.
- the name of the database that will be used by ZenML. The database does not
have to exist prior to the deployment (ZenML will create it on first start).
However, you need to create the database if you follow the best practice of
restricting database user privileges to only access it.
- if you plan on using SSL to secure the client database connection, you may
also need to prepare additional SSL certificates and keys:

  - the TLS CA certificate that was used to sign the server TLS certificate, if
    you're using a self-signed certificate or signed by a custom certificate
    authority that is not already trusted by default by most operating systems. 

>   - the TLS client certificate and key. Only needed if you decided to use

Reviewer: ```suggestion
  - the TLS client certificate and key. This is only needed if you decide to use
```

    client certificates for your DB connection (some managed DB services support
    this, CloudSQL is an example).

### Optional cluster services

> It is common practice to install additional infrastructure related services in

Reviewer: ```suggestion
It is common practice to install additional infrastructure-related services in
```

a Kubernetes cluster to support the deployment and long-term management of
applications. For example:

- an Ingress service like [nginx-ingress](https://kubernetes.github.io/ingress-nginx/deploy/)
is recommended if you want to expose HTTP services to the internet. An Ingress

> is required if you want to use secure HTTPs for your ZenML deployment. The

Reviewer: Is this the plural of HTTP, or HTTPS?

Author: hehe, should indeed be HTTPS :)

Author: done

alternative is to use a LoadBalancer service to expose the ZenML service using
plain HTTP, but this is not recommended for production.

> - [cert-manager](https://cert-manager.io/docs/installation/) is

Reviewer: ```suggestion
- a [cert-manager](https://cert-manager.io/docs/installation/) is
```

recommended if you want to generate and manage TLS certificates for your ZenML
deployment. It can be used to automatically provision TLS certificates from a
certificate authority (CA) of your choice, such as [Let's Encrypt](https://letsencrypt.org/).
As an alternative, the ZenML Helm chart can be configured to auto-generate
self-signed or you can generate the certificates yourself and provide them to
the Helm chart, but this makes it more difficult to manage the certificates
and you need to manually renew them when they expire.

## ZenML Helm Installation

### Configure the Helm chart

To customize the Helm chart for your deployment, you should create a copy of
the `values.yaml` file located in the `src/zenml/zen_server/deploy/helm` folder
(let’s call this `custom-values.yaml`). You’ll use this as a template to
customize your configuration. Any values that you don’t override you should
simply remove from your `custom-values.yaml` file to keep it clean and

> compatible with future helm chart releases.

Reviewer: ```suggestion
compatible with future Helm chart releases.
```
In general, there's a mix of `helm` and `Helm` used in this file. Probably should standardise.

In most cases, you’ll need to change the following configuration values in
`custom-values.yaml`:

- default username and password values
- the database configuration, if you mean to use an external database:

    - the database URL, formatted as
    `mysql://<username>:<password>@<hostname>:<port>/<database>`
    - CA and/or client TLS certificates, if you’re using SSL to secure the
    connection to the database

> - the ingress configuration, if enabled:

Reviewer: ```suggestion
- the Ingress configuration, if enabled:
```
This was capitalised above. Unsure if it also needs the same treatment here?

    - enabling TLS
    - enabling self-signed certificates
    - configuring the hostname that will be used to access the ZenML server,

>     if different than the IP address or hostname associated with the Ingress

Reviewer: ```suggestion
    if different from the IP address or hostname associated with the Ingress
```
